maliput/src/math/mod.rs

```diff
@@ -0,0 +1,154 @@
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/// A 3D vector.
+/// Wrapper around C++ implementation `maliput::math::Vector3`.
+///
+/// ## Example
+///
+/// ```rust, no_run
+/// use maliput::math::Vector3;
+///
+/// fn main() {
+///     let v = Vector3::new(1.0, 2.0, 3.0);
+///     println!("v = {}", v);
+///     assert_eq!(v.x(), 1.0);
+///     assert_eq!(v.y(), 2.0);
+///     assert_eq!(v.z(), 3.0);
+///     assert_eq!(v.norm(), (1.0_f64.powi(2) + 2.0_f64.powi(2) + 3.0_f64.powi(2)).sqrt());
+///     assert_eq!(v.dot(&v), 1.0_f64.powi(2) + 2.0_f64.powi(2) + 3.0_f64.powi(2));
+///     assert_eq!(v.cross(&v), Vector3::new(0.0, 0.0, 0.0));
+/// }
+/// ```
+pub struct Vector3 {
+    v: cxx::UniquePtr<maliput_sys::math::ffi::Vector3>,
+}
+
+impl Vector3 {
+    /// Create a new `Vector3` with the given `x`, `y`, and `z` components.
+    pub fn new(x: f64, y: f64, z: f64) -> Vector3 {
+        Vector3 {
+            v: maliput_sys::math::ffi::Vector3_new(x, y, z)
+        }
+    }
+    /// Get the `x` component of the `Vector3`.
+    pub fn x(&self) -> f64 {
+        self.v.x()
+    }
+    /// Get the `y` component of the `Vector3`.
+    pub fn y(&self) -> f64 {
+        self.v.y()
+    }
+    /// Get the `z` component of the `Vector3`.
+    pub fn z(&self) -> f64 {
+        self.v.z()
+    }
+    /// Get the norm of the `Vector3`.
+    pub fn norm(&self) -> f64 {
+        maliput_sys::math::ffi::Vector3_norm(&self.v)
+    }
+    /// Get the dot product of the `Vector3` with another `Vector3`.
+    pub fn dot(&self, w: &Vector3) -> f64 {
+        maliput_sys::math::ffi::Vector3_dot(&self.v, &w.v)
+    }
+    /// Get the cross product of the `Vector3` with another `Vector3`.
+    pub fn cross(&self, w: &Vector3) -> Vector3 {
+        Vector3 {
+            v: maliput_sys::math::ffi::Vector3_cross(&self.v, &w.v)
+        }
+    }
+}
+
+impl PartialEq for Vector3 {
+    fn eq(&self, other: &Self) -> bool {
+        maliput_sys::math::ffi::Vector3_equals(&self.v, &other.v)
+    }
+
+}
+
+impl Eq for Vector3 {}
+
+impl std::fmt::Display for Vector3 {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", maliput_sys::math::ffi::Vector3_to_str(&self.v))
+    }
+}
+
+impl std::fmt::Debug for Vector3 {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_struct("Vector3")
+            .field("x", &self.x())
+            .field("y", &self.y())
+            .field("z", &self.z())
+            .finish()
+    }
+}
+
+mod tests {
```

Comment:
Is it common to keep tests within the same file than the type implementation?

Reply:
Are we missing debug and non debug format checks?

Reply:
> Is it common to keep tests within the same file than the type implementation?

Yes. Unit tests should be located at the same file or evean you can create a submodule for tests.

Integration Tests are the ones that are expected to be put under /root/tests folder

See:
 - https://doc.rust-lang.org/rust-by-example/testing.html
 - https://users.rust-lang.org/t/should-unit-tests-really-be-put-in-the-same-file-as-the-source/62153

Reply:
> Are we missing debug and non debug format checks?

Not sure if I follow you, can you explain?